src/moonlink/src/storage/mooncake_table/persistence_buffer.rs

```diff
@@ -88,15 +88,25 @@ impl UnpersistedRecords {
 
     /// Get unpersisted data files as hash set for lookup.
     pub(crate) fn get_unpersisted_data_files_set(&self) -> HashSet<MooncakeDataFileRef> {
-        self.new_data_files.iter().cloned().collect::<HashSet<_>>()
+        let expected_len = self.new_data_files.len() + self.compacted_data_files_to_add.len();
+        let mut unpersisted_data_files = HashSet::new();
+        unpersisted_data_files.extend(self.new_data_files.iter().cloned());
+        unpersisted_data_files.extend(self.compacted_data_files_to_add.iter().cloned());
+        assert_eq!(expected_len, unpersisted_data_files.len());
+        unpersisted_data_files
     }
     /// Get unpersisted file indices as hash set for lookup.
     #[allow(clippy::mutable_key_type)]
     pub(crate) fn get_unpersisted_file_indices_set(&self) -> HashSet<GlobalIndex> {
-        self.new_file_indices
-            .iter()
-            .cloned()
-            .collect::<HashSet<_>>()
+        let expected_len = self.new_file_indices.len()
```

Comment:
This was a bug: it's used to decide which file indices could be used to perform index merge.
Currently index merge is only attempted upon persisted ones, so need to consider all unpersisted items; I missed index merge and compacted results.